geode-core/src/main/java/org/apache/geode/management/internal/operation/OperationHistoryManager.java

```diff
@@ -27,10 +28,16 @@
 import org.apache.geode.management.api.ClusterManagementOperation;
 import org.apache.geode.management.api.JsonSerializable;
 
+/**
+ * Retains references to all running and some recently-completed operations.
+ *
+ * The policy for expiring completed operations is subject to change, but may be based on age,
+ * count, when it was last accessed, or some combination thereof.
+ */
 @Experimental
 public class OperationHistoryManager {
-  private final ConcurrentMap<String, CompletableFuture> inProgressHistory;
-  private final Map<String, CompletableFuture> completedHistory;
+  private final ConcurrentMap<String, OperationInstance> inProgressHistory;
```

Comment:
what's the reason that we have to use two maps to hold the history?

Reply:
One keeps track of any commands still in progress (inProgessHistory). The other is limited to a certain number of completed commands (completedHIstory) and is implemented with an LRUMap that only remembers a certain number.

Reply:
So it is to guarantee that an "in progress" operation would not get cleaned out if it gets stale? what if an operation "never" finishes somehow?

Reply:
If it is still running in the locator then it will still be in the inProgress map. Can you define what you mean by "stale"? If it is still running then we already have a bunch of other resources allocated for it (like a jvm thread) so I don't think have any entry in a map is a problem. Once it completes in the locator it will move to the completed map which has a limit on how many things can be in it. It does not need a client to be checking the status for this move to happen.

Reply:
is it ever possible that an operation instance would stay in this "in progress" map for ever? like that thread just dies or that operation never completes...

Reply:
The only way I know for a thread to "die" is by throwing some `Throwable`.  We execute using `CompletableFuture.supplyAsync` which completes exceptionally if it catches Throwable.  On the other hand, if you're talking about an operation that hangs, it seems like it-progress forever is an accurate depiction of that state.  Are you requesting that we put in a global or per-operation-type timeout value beyond which any operations still in progress will be forcibly killed?

Reply:
No, I am just wondering why we need to use two maps for these operationInstances, can we use just one LRU map to keep all of them? So if one operation is "in-progress" forever, it can still be wiped out eventually.

Reply:
Because we would like to guarantee that status will not disappear before the operation has a chance to complete.  If you use a single LRU cache with size=100, then start 101 async operations, you are guaranteeing that you'll never be able to retrieve status for one of them.  Although we may still have then problem once they complete, you would still be guaranteed to lose one result.  Maybe this should be reworked to a single map with a time-based expiration policy e.g. all results are kept until 1 hour after operation completes?

Reply:
changed to 2-hour retention policy of completed operations and a single map rather than two separate ones